### PR TITLE
Changes to restore ipsec show command

### PR DESCRIPTION
**Pull request by bradyallenjohnson**
- Now called "ipsec briefconnectionstatus", and is a simpler version "ipsec connectionstatus"
- Now implemented in c-code in the whack binary

### CODE REVIEW

**Comment by cagney** [comment]
It looks like the output is the same as:
```
ipsec connectionstatus | grep '[.][.][.]'
```

I wonder what we would think of displaying if the command was called `ipsec routestatus`

**Comment by bradyallenjohnson** [comment]
> It looks like the output is the same as:
> 
> ```
> ipsec connectionstatus | grep '[.][.][.]'
> ```
> 
> I wonder what we would think of displaying if the command was called `ipsec routestatus`

I show the output in this email:
https://lists.libreswan.org/pipermail/swan-dev/2023-October/004654.html

**Comment by paulwouters** [comment]
Personally, I strongly prefer the first two things are source/mask <=> dest/mask,
as that is the core info people usually want to see/confirm. Having this
split in two makes this much harder to read. If you want to avoid adding
up the counters, perhaps this format could be considered:

     000 172.16.20.0/24 <=> 172.16.10.0/24 from 172.22.18.102 to 172.22.18.101 (2KiB/1KiB) vpnclient.gwn02.xyz.com, reqid=16388

Thanks for adding reqid, that is very useful.
Perhaps peer ID should be the last entry and reqid should come first, as those peer IDs can be very long when certificates are used.

**Comment by cagney** [comment]
Can the addition of `reqid` to `ipsec connectionstatus` please be moved to a separate patch - it will make a mess of testing so better to isolate it.

**Comment by bleve** [comment]
Now handling multiple traffic-selectors is problematic. Current listing:

```
000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in)        foo-bar-v4, reqid=16440
000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)   foo-bar-v4, reqid=16440
000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in) foo-bar-v4, reqid=16440
000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)    foo-bar-v4, reqid=16440

```

I suggest something like this, Output changing also other direction traffic to out.

`
000 193.65.3.96/27,172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27,172.23.128.0/23 @ 188.127.201.217 (4KiB out) foo-bar-v4, reqid=16440
`

**Comment by letoams** [comment]
On Thu, 2 Nov 2023, cagney wrote:

> Subject: Re: [libreswan/libreswan] Changes to restore ipsec show command (PR

> Can the addition of reqid to ipsec connectionstatus please be moved to a separate patch - it will make a mess of
> testing so better to isolate it.

We didn't use "ipsec show" in tests anymore. We had removed it. I don't
think any of this would affect any tests?

**Comment by letoams** [comment]
On Thu, 2 Nov 2023, Tuomo Soini wrote:

> Now handling multiple traffic-selectors is problematic. Current listing:
> 
> 000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in)        foo-bar-v4,
>  reqid=16440
> 000 193.65.3.96/27 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)   foo-bar-v4, reqid=1
> 6440
> 000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  188.127.201.224/27 @ 188.127.201.217 (4KiB in) foo-bar-v4, reqid=1
> 6440
> 000 172.23.0.0/17 @ 193.65.2.90 (5KiB in)  <==>  172.23.128.0/23 @ 188.127.201.217 (4KiB in)    foo-bar-v4, reqid=1
> 6440

Other than traffic counters, I think this is good.

> 000 193.65.3.96/27,172.23.0.0/17 @ 193.65.2.90 (5KiB in) <==> 188.127.201.224/27,172.23.128.0/23 @ 188.127.201.217
> (4KiB out) foo-bar-v4, reqid=16440

With more subnets this again only becomes machine parseable to read. The
idea of "ipsec show" is that humans can read it easilly.

**Comment by cagney** [comment]
Once the command's name is settled we can move the code to `whack_<command>.[hc]` (this can happen post merge).

**Comment by cagney** [comment]
Can I suggest dropping the header / footer - keep it brief.

Was each line going to be prefixed with the connection name?  Without this it is hard to tie the information back to the configuration.

Can I suggest including the routing so it is clear which are established, or are only established connections shown like for `ipsec trafficstatus`.

Is the name settled? `ipsec briefconnectionstatus` is a mouth full (I suggested routestatus before).

Having one SPD per line will make it easier to understand (but confusing when it comes to traffic counts as they are not per line), think about:
```
"my-connection": 192.0.1.23 ... 192.0.1.45 4kb in 8kb out routed-tunnel
"my-connection":   192.0.2.0/24 === 192.0.3.0/24 reqid 1
"my-connection":   1::/64 === 2:/64 reqid 2
```

**Comment by letoams** [comment]
On Thu, 2 Nov 2023, cagney wrote:

> Can I suggest dropping the header / footer - keep it brief.

Yes.

> Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to
> the configuration.

I hope not. I think the idea is to closely keep with the old "ipsec
eroute" quick overview idea.

> Can I suggest including the routing so it is clear which are established, or are only established connections shown
> like for ipsec trafficstatus.

I think only established ones are?

> Is the name settled? ipsec briefconnectionstatus is a mouth full (I suggested routestatus before).

"ipsec eroute" ? :)

**Comment by cagney** [comment]
> @cagney wrote
> > Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to the configuration.

@letoams wrote
> I hope not. I think the idea is to closely keep with the old "ipsec eroute" quick overview idea.

Then I'm struggling to see the advantage of this vs `ipsec show` and the host specific equivalents `ip xfrm`, `setkey`, et.al.

When I run `ipsec trafficstatus` and see a problem, I can follow it with `ipsec connectionstatus CONNECTION` (#1308) and examine the details.  Here I can't.

**Comment by bradyallenjohnson** [comment]
> > @cagney wrote
> > > Was each line going to be prefixed with the connection name? Without this it is hard to tie the information back to the configuration.
> 
> @letoams wrote
> 
> > I hope not. I think the idea is to closely keep with the old "ipsec eroute" quick overview idea.
> 
> Then I'm struggling to see the advantage of this vs `ipsec show` and the host specific equivalents `ip xfrm`, `setkey`, et.al.
> 
> When I run `ipsec trafficstatus` and see a problem, I can follow it with `ipsec connectionstatus CONNECTION` (#1308) and examine the details. Here I can't.

Yes, the configuration name is indeed displayed in "ipsec connectionstatus" and in the new "ipsec briefconnectionstatus" command.
In this output, the configuration name (which is also the connection name) is vpnclient.gwn02.xyz.com:

000 [172.16.20.0/24](http://172.16.20.0/24) @ 172.22.18.102 (2KiB in)  <==>  [172.16.10.0/24](http://172.16.10.0/24) @ 172.22.18.101 (1KiB in) vpnclient.gwn02.xyz.com, reqid=16388

**Comment by cagney** [comment]
@bradyallenjohnson wrote
> Yes, the configuration name is indeed displayed [...] In this output, the configuration name (which is also the connection name) is vpnclient.gwn02.xyz.com:
```
000 [172.16.20.0/24](http://172.16.20.0/24) @ 172.22.18.102 (2KiB in) <==> [172.16.10.0/24](http://172.16.10.0/24) @ 172.22.18.101 (1KiB in) vpnclient.gwn02.xyz.com, reqid=16388
```
ah, so that is what that is, good!  I had assumed it was some DNS info or some such and ignored it.  It should be displayed using jam_connection(), I added a note.

File: programs/pluto/whack_connectionstatus.c [hunk 1]
```diff
@@ -325,7 +329,7 @@ static void show_one_spd(struct show *s,
 		ip_address their_sourceip = spd_end_sourceip(spd->remote);
 		jam_string(buf, " their_ip=");
 		OPT_HOST(their_sourceip);
-		jam_string(buf, ";");
+		jam(buf, "; reqid=%u;", c->child.reqid);
```
[the remark this answers is not in the document]
Comment (bradyallenjohnson) [reply programs/pluto/whack_connectionstatus.c:332]: Ok, I'll remove this from this PR.

File: programs/pluto/whack_connectionstatus.c [hunk 2]
```diff
@@ -934,3 +938,156 @@ void whack_connectionstatus(const struct whack_message *m, struct show *s)
...
+
+		jam_string(buf, "\tfrom ");
+		jam_end_host(buf, c, spd->local->host);
+		jam_string(buf, " to ");
+		jam_end_host(buf, c, spd->remote->host);
+		jam_humber_uintmax(buf, " (", get_child_bytes(c, DIRECTION_INBOUND), "B");
+		jam_humber_uintmax(buf, "/", get_child_bytes(c, DIRECTION_OUTBOUND), "B)");
+
+		// name and reqid
+		jam(buf, "\t"PRI_CONNECTION", reqid=%u", c->name, instance, c->child.reqid);
```
[the remark this answers is not in the document]
Comment (bradyallenjohnson) [reply programs/pluto/whack_connectionstatus.c:1020]: Ok, I will change this.

File: programs/pluto/whack_connectionstatus.h [hunk 3]
```diff
@@ -20,6 +20,7 @@ struct whack_message;
 struct show;
 
 void whack_connectionstatus(const struct whack_message *m, struct show *s);
+void whack_brief_connectionstatus(const struct whack_message *m, struct show *s);
```
[the remark this answers is not in the document]
Comment (bradyallenjohnson) [reply programs/pluto/whack_connectionstatus.h:23]: Ok

File: programs/pluto/whack_connectionstatus.c [hunk 4]
```diff
@@ -46,7 +46,11 @@
 #include "log.h"
 #include "show.h"
 #include "crypto.h"		/* for show_ike_alg_connection() */
-#include "plutoalg.h"		/* for show_kernel_alg_connection() */
+#include "plutoalg.h"	/* for show_kernel_alg_connection() */
+#include "kernel.h"		/* for enum direction */
+
+/* Passed in to jam_end_client() */
+static const char END_SEPARATOR[] = "===";
```
[the remark this answers is not in the document]
Comment (bradyallenjohnson) [reply programs/pluto/whack_connectionstatus.c:53]: The END_SEPARATOR is indeed used in the calls to jam_end_client(). The jam_end_client() function is called with the separator parameter set to either NULL or END_SEPARATOR.

[the remark this answers is not in the document]
Comment (cagney) [reply programs/pluto/whack_connectionstatus.c:53]: Yes, it's treated like a boolean though
```
	if (left_right == RIGHT_END && separator != NULL) {
		jam_string(buf, "===");
	}
```

[the remark this answers is not in the document]
Comment (bradyallenjohnson) [reply programs/pluto/whack_connectionstatus.c:53]: Oh, I see, I'll fix that, nice catch.

**Comment by bleve** [comment]
I don't see any need for output prefix and suffix parts, I suggest removing those.

**Comment by bleve** [comment]
And I still don't like this:
```

000 193.65.3.96/27 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 193.65.3.96/27 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 172.23.0.0/17 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
000 172.23.0.0/17 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440

```

When we could do this all in one line.

```
000 193.65.3.96/27,172.23.0.0/17 <==> 188.127.201.224/27,172.23.128.0/23        from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)     "foo-bar-v4", reqid=16440
```

**Comment by bradyallenjohnson** [comment]
> I don't see any need for output prefix and suffix parts, I suggest removing those.

Which prefix/suffix parts are you referring to?

**Comment by bleve** [comment]
> > I don't see any need for output prefix and suffix parts, I suggest removing those.
> 
> Which prefix/suffix parts are you referring to?

```
000 Connection list:
000  
```

And

```
000  
000 Total IPsec connections: loaded 38, active 19
```

**Comment by bradyallenjohnson** [comment]
> > > I don't see any need for output prefix and suffix parts, I suggest removing those.
> > 
> > 
> > Which prefix/suffix parts are you referring to?
> 
> ```
> 000 Connection list:
> 000  
> ```
> 
> And
> 
> ```
> 000  
> 000 Total IPsec connections: loaded 38, active 19
> ```

The final line seems very useful to me. If there are more than 4 or 5 connections, I would like to have a simple count of the number of connections.

**Comment by bradyallenjohnson** [comment]
> And I still don't like this:
> 
> ```
> 
> 000 193.65.3.96/27 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 193.65.3.96/27 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 172.23.0.0/17 <==> 188.127.201.224/27	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> 000 172.23.0.0/17 <==> 172.23.128.0/23	from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)	"foo-bar-v4", reqid=16440
> ```
> 
> When we could do this all in one line.
> 
> ```
> 000 193.65.3.96/27,172.23.0.0/17 <==> 188.127.201.224/27,172.23.128.0/23        from 193.65.2.90 to 188.127.201.217 (119KiB/112KiB)     "foo-bar-v4", reqid=16440
> ```

I disagree. The second output is much harder to read.

**Comment by bleve** [comment]
We have `ipsec briefstatus` for that.

**Comment by cagney** [comment]
Having one SPD per line is consistent with the old `ipsec show` command and makes parsing the output using scripts much easier.  This isn't a log file where we're trying to squeeze as much as possible on a single line.

**Comment by cagney** [comment]
I don't think the output should be decorated with the header/footer as it, again, makes it more consistent with the old `ipsec show`, however ...

`ipsec briefconnectionstatus ; ipsec briefstatus` is *not* equivalent to including the header footer but `ipsec whack --briefconnectionstatus --briefstatus` kind of is (yea, long standing quirk in how this works).

**Comment by cagney** [comment]
@bradyallenjohnson  I still see
```
@@ -325,7 +329,7 @@ static void show_one_spd(struct show *s,
                ip_address their_sourceip = spd_end_sourceip(spd->remote);
                jam_string(buf, " their_ip=");
                OPT_HOST(their_sourceip);
-               jam_string(buf, ";");
+               jam(buf, "; reqid=%u;", c->child.reqid);
 #undef OPT_HOST
        }
 
```

**Comment by bradyallenjohnson** [comment]
> jam_string(buf, ";");

Rebased and fixed in the latest push.

**Comment by cagney** [comment]
I've merged and shuffled things so it is easier to look at and play with.  And added your (C).

What happens in transport mode when host==client?  It looks like jam_end_client() prints nothing.
We still lack documentation, tests, and consensus on the final output.

**Comment by bradyallenjohnson** [comment]
> I've merged and shuffled things so it is easier to look at and play with. And added your (C).
> 
> What happens in transport mode when host==client? It looks like jam_end_client() prints nothing. We still lack documentation, tests, and consensus on the final output.

There are 3 cases affected by this: transport mode when host==client, tunnel mode with host-to-subnet, and tunnel mode with subnet-to-host.

I propose the following output for these 3 cases:

transport mode when host == client for both the local and remote:
000 host <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

tunnel mode with host-to-subnet:
000 host <==> 172.16.10.0/24  from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

tunnel mode with subnet-to-host:
000 172.16.20.0/24 <==> host  from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388

**Comment by letoams** [comment]
On Mon, 20 Nov 2023, Brady Johnson wrote:

> There are 3 cases affected by this: transport mode when host==client, tunnel mode with host-to-subnet,
> and tunnel mode with subnet-to-host.
> 
> I propose the following output for these 3 cases:
> 
> transport mode when host == client for both the local and remote:
> 000 host <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun", reqid=16388
> 
> tunnel mode with host-to-subnet:
> 000 host <==> 172.16.10.0/24 from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun",
> reqid=16388
> 
> tunnel mode with subnet-to-host:
> 000 172.16.20.0/24 <==> host from 172.22.18.102 to 172.22.18.101 (0B/0B) "gwn02_transport_tun",
> reqid=16388

Instead of the string "host" I would just repeat the IP/mask of the
host instead. This will make it easier on tools parsing the output
automatically - they wouldn't really need to know if it was tunnel
or transport or net-net or host-net/net-host.

Paul

**Comment by bradyallenjohnson** [comment]
Ok, Paul, I'll do that and push the changes.

**Comment by bradyallenjohnson** [comment]
I just pushed changes to handle the 4 use cases of host2host (transport mode), host2subnet, subnet2host, and subnet2subnet with the output suggested by @letoams as follows:

000 172.22.18.102 <==> 172.22.18.101	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_transport_tun", reqid=16388

000 172.22.18.102 <==> 172.16.10.0/24	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_host2subnet_tun", reqid=16392

000 172.16.20.0/24 <==> 172.22.18.101	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"gwn02_subnet2host_tun", reqid=16388

000 172.16.20.0/24 <==> 172.16.10.0/24	from 172.22.18.102 to 172.22.18.101 (0B/0B)	"vpnclient.gwn02.xyz.com", reqid=16396

@cagney Can you point out where the documentation for this is so I can update it. Also, for the tests, is the libreswan/testing/pluto/whack-02-globalstatus test a good example to follow?

**Comment by cagney** [comment]
I merged.

I do wonder if the jam_end_child() calls can be reduced to jam_selector() and jam_end_host() to jam_address() when showing the SPD.  BTW, like how it shows `172.22.18.102` (i.e., address) and not `172.22.18.102/32` (i.e., selector).

For documentation, it's in programs/ipsec/ipsec_<cmd>.8.xml, I'm not sure how @bleve is keeping it all co-ordinated.

For tests, tricky, I've never found a good basic template (the basic tests often aren't that basic).  Perhaps `testing/pluto/ikev2-05-basic-psk`  Since you've figured out the configs, you could tweak that to up/show/down each combination.

**Comment by cagney** [comment]
Moved to 5.1 as not a blocker.  Since `whack_briefconnectionstatus.[hc]` is nice an isolated we can, in truth, still pull stuff into 5.0

**Comment by bleve** [comment]
This pull request was already merged in, closing. For more changes, create a new one.

**Pull request closed without merging**